ratpack-exec/src/main/java/ratpack/exec/util/PromiseZipper.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ratpack.exec.util;
+
+import ratpack.exec.ExecResult;
+import ratpack.exec.Promise;
+import ratpack.exec.util.internal.DefaultPromiseZipper;
+import ratpack.func.*;
+
+public interface PromiseZipper {
+
+
+  <T1, T2, R> Promise<R> yield(BiFunction<T1, T2, R> function);
```

Comment:
The `yield` and `yieldAll` could move into a `PromiseZipper2` interface to keep things consistent. Then `PromiseZipper` would be focused on `zip`.

Also, each `PromiseZipper*` interface could inherit another interface that provides an `execInit` function. Then `DefaultPromiseZipper` would use it on each forked execution.

Something like:
```
PromiseZipper zipper = Promises.zip(p1, p2, p3);
zipper.execInit(e -> e.add(Execution.current().get(String.class));
Promise<String> result = zipper.yield((r1, r2, r3) -> r1.val + r2.val);
```